site/src/components/PopoverContainer/PopoverContainer.tsx

```diff
@@ -0,0 +1,149 @@
+/**
+ * @file Abstracts over MUI's Popover component to simplify using it (and hide)
+ * some of the wonkier parts of the API.
+ *
+ * Just place a button and some content in the component, and things just work.
+ * No setup needed with hooks or refs.
+ */
+import {
+  type KeyboardEvent,
+  type ReactElement,
+  useEffect,
+  useRef,
+  useState,
+  PropsWithChildren,
+} from "react";
+
+import { type Theme, type SystemStyleObject } from "@mui/system";
+import Popover, { type PopoverOrigin } from "@mui/material/Popover";
+
+type Props = PropsWithChildren<{
+  /**
+   * Does not require any hooks or refs to work. Also does not override any refs
+   * or event handlers attached to the button.
+   */
+  anchorButton: ReactElement;
+  width?: number;
+  originX?: PopoverOrigin["horizontal"];
+  originY?: PopoverOrigin["vertical"];
+  sx?: SystemStyleObject<Theme>;
+}>;
+
+function getButton(container: HTMLElement) {
+  return (
+    container.querySelector("button") ??
+    container.querySelector('[aria-role="button"]')
+  );
+}
+
+export function PopoverContainer({
+  children,
+  anchorButton,
+  originX = 0,
+  originY = 0,
+  width = 320,
+  sx = {},
+}: Props) {
+  const buttonContainerRef = useRef<HTMLDivElement>(null);
+
+  // Ref value is for effects and event listeners; state value is for React
+  // renders. Have to duplicate state because after the initial render, it's
+  // never safe to reference ref contents inside a render path, especially with
```

Comment:
Because React never rerenders the component? Why is particularly the case with React 18?

Reply:
So technically it's a concern with all versions of React that use hooks – the React team really wants you to treat refs like black boxes during renders. As in, the ref itself is fine – it's just a box for storing values, and you can work with the box and attach it to elements. But the only times it's safe to peak into the box to read from it or update the contents are one of two situations:

1. On the very first render when you're setting up the ref for the first time. For example:
```ts
// This is only safe if the other effects and event handlers can't change the ref to null
const abortControllerRef = useRef<AbortController>(null!);
if (abortController.current === null) {
  abortControllerRef.current = new AbortController();  
}
```
2. Inside the various useEffect hooks or event handlers

Reply:
But it's more of a problem with React 18's concurrency mode, which has the ability to stop/throw away, de-prioritize, and redo renders. They can only do that because renders are supposed to be pure, and reading from/writing refs is technically a side effect/"side cause", so there's a risk that as their render behavior changes (which might happen soon with React Forget), there's more risks of things breaking if the rules aren't followed 100%

Reply:
> They can only do that because renders are supposed to be pure, and reading from/writing refs is technically a side effect/"side cause", so there's a risk that as their render behavior changes (which might happen soon with React Forget), there's more risks of things breaking if the rules aren't followed 100%

Ahhh, makes sense. Thanks for the explanation!